### PR TITLE
Unit testing for fact.js

### PR DESCRIPTION
**Pull request by kathyn262**
This pull request includes a

- [ ] Bug fix
- [x] New feature
- [x] Testing
- [ ] Translation

The following changes were made

- Added __ tests __ folder with subfolder unit 
- Added fact.test.js into unit subfolder
- Exported functions in fact.js
- Added tests in fact.test.js

Comments
- Only tested existing functions 
- Excluded testing for error handling as there weren't many error handlers 
- Noticed error handling needed for the function getRandomApiNum in fact.js
  - When number is not in range of min and max, undefined is returned instead of the number. For example, {min: 2010, max: 2011} returns undefined rather than the number itself. "undefined is an unremarkable number." This only occurs with types date and trivia. 
  - When min and max are invalid values, undefined is also returned. For example, {min: 10, max: -20} returns undefined instead of an error message. This occurs with all 4 types (date, trivia, math, year). 
 

#5

### CODE REVIEW

**Comment by mmmaaatttttt** [comment]
@silasburger @maxnawa31 @Ronihe do one of you want to take a first pass at this?

**Comment by kathyn262** [comment]
Let us know if there's any way for us to make it easier for reviewers! :)

**Comment by mmmaaatttttt** [comment]
Hey @kathyn262 - sorry for the delay. i'll take the lead on this review next week!

File: __tests__/unit/fact.test.js [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+    expect(sentence).toBe(
+      "January 1st is the day that no newsworthy events happened."
+    );
+  });
+});
+
+describe("dataPairs", () => {
+  // dataPairs returns
+  // {"date": [Array], "math": [Array], "trivia": [Array], "year": [Array]}
+  // {'number': 1, string: '1'}
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:227]: this part is a bit confusing, does it return an object with 4 keys or 2?

File: __tests__/unit/fact.test.js [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+  test("return object with keys 'date', 'math', 'trivia', 'year'", function () {
+    expect(Object.keys(dataPairs)).toEqual(["date", "year", "trivia", "math"]);
+  });
+
+  test("date contains data in correct format", function () {
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:233]: can you add a comment explaining what's going on here? it's not super clear to me what this function is doing, or why all of the nested arrays look like they have `{ number: 1, string: "1" }` in them. are there other things in there too? is that why you're using `arrayContaining` / `objectContaining`?

File: __tests__/unit/fact.test.js [hunk 3]
```diff
@@ -0,0 +1,359 @@
...
+    expect(emptyWhitelist).toEqual({});
+  });
+
+  test("return empty object when whitelist key is not in object", function () {
+    let notPresentKey = filterObj(
+      {
+        hello: "world",
+        not: "this",
+      },
+      ["whiskey"]
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:299]: yesssss!

File: __tests__/unit/fact.test.js [hunk 4]
```diff
@@ -0,0 +1,359 @@
...
+describe("apiExtend()", () => {
+  test("filters object and extends with new data", function () {
+    let extend = apiExtend({ year: 2020, random: "lol" }, { newData: "wow" });
+    expect(extend).toEqual({ year: 2020, newData: "wow" });
+  });
+});
+
+describe("getFact()", () => {
+  test("return error object for invalid input type", function () {
+    let invalidType = getFact(1000, "sdasd", {});
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:314]: here too, would be nice if this call looked like:
```suggestion
    let invalidType = getFact({ number: 1000, type: "sdasd" });
```

**Review by mmmaaatttttt — CHANGES_REQUESTED** [review]
@kathyn262 thanks for your patience! Had a chance to review. I think we should use this as an opportunity to refactor some of the functions you're testing. overall things look good.

Also, can you please remove `models/year/raw/.DS_Store` from the PR? Looks like we may need to add `.DS_Store` to the `.gitignore` as well.

**Comment by WinnieGotYu** [comment]
@mmmaaatttttt  @kathyn262  and I have completed all the suggested changes from above. We're ready for a review. Thank you!

File: __tests__/unit/fact.test.js [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+  test("return random number when min and max are undefined", function () {
+    let noMinMax = getRandomApiNum({ type: "year" });
+    expect(typeof noMinMax).toBe("number");
+  });
+});
+
+describe("getSentence() for type 'year'", () => {
+  test("When wantFragment is true, returns partial text only", function () {
+    const sentence = getSentence({
+      wantFragment: { fragment: true },
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:114]: this looks a bit odd, can you make it so that it just looks like:

```suggestion
      wantFragment: true,
```

File: __tests__/unit/fact.test.js [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+      },
+    });
+    expect(sentence).toBe(
+      "Sweyn I establishes Danish control over part of Norway"
+    );
+  });
+
+  test("When wantFragment is undefined, returns full sentence", function () {
+    const sentence = getSentence({
+      wantFragment: undefined,
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:128]: you should be able to omit `wantFragment` entirely if you want its value to be `undefined`

File: __tests__/unit/fact.test.js [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+      },
+    });
+    expect(sentence).toBe(
+      "1000 is the year that Sweyn I establishes Danish control over part of Norway."
+    );
+  });
+
+  test("When data has key of date, returns full sentence with date in the end", function () {
+    const sentence = getSentence({
+      wantFragment: undefined,
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:142]: same here, can you remove this line?

File: __tests__/unit/fact.test.js [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+      },
+    });
+    expect(sentence).toBe(
+      "1000 is the year that Stephen I becomes King of Hungary, which is established as a Christian kingdom on December 25th."
+    );
+  });
+
+  test("Return sentence when year is invalid", function () {
+    const sentence = getSentence({
+      wantFragment: undefined,
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:158]: remove

File: __tests__/unit/fact.test.js [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+      "10000000000 is the year that nothing remarkable happened."
+    );
+  });
+});
+
+describe("getSentence() for types: trivia and math", () => {
+  // trivia and math share the same prefix
+  test("When wantFragment is true, returns partial text only", function () {
+    const sentence = getSentence({
+      wantFragment: { fragment: true },
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:175]:
```suggestion
      wantFragment: true,
```

File: __tests__/unit/fact.test.js [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+      },
+    });
+    expect(sentence).toBe(
+      "the number of keys on a piano (36 black and 52 white)"
+    );
+  });
+
+  test("When wantFragment is undefined, returns full sentence", function () {
+    const sentence = getSentence({
+      wantFragment: undefined,
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:189]: remove

File: __tests__/unit/fact.test.js [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+      },
+    });
+    expect(sentence).toBe(
+      "88 is the number of keys on a piano (36 black and 52 white)."
+    );
+  });
+
+  test("When trivia number is invalid, returns sentence", function () {
+    const sentence = getSentence({
+      wantFragment: undefined,
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:203]: remove

File: __tests__/unit/fact.test.js [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+      },
+    });
+    expect(sentence).toBe("88888 is an unremarkable number.");
+  });
+});
+
+describe("getSentence() for type 'date' ", function () {
+  test("When wantFragment is undefined, returns full sentence", function () {
+    const sentence = getSentence({
+      wantFragment: undefined,
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:217]: remove

File: __tests__/unit/fact.test.js [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+      number: 1 / 1,
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:218]: why is this `1 / 1` instead of just `1`? if there's a reason, can you add a comment giving it? if there's no reason, let's just use `1`

File: __tests__/unit/fact.test.js [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+    const yearMsgs = [
+      "9999999999 is the year that nothing remarkable happened.",
+      "9999999999 is the year that the Earth probably went around the Sun.",
+      "9999999999 is the year that nothing interesting came to pass.",
+      "9999999999 is the year that we do not know what happened.",
+    ];
+    const sentence = getDefaultMsg({
+      number: 9999999999,
+      type: "year",
+      options: {},
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:242]: would be nice if you could just not provide `options` at all if you didn't want to pass any.

File: __tests__/unit/fact.test.js [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+    const mathMsgs = [
+      "9999999999 is an uninteresting number.",
+      "9999999999 is a boring number.",
+      "9999999999 is an unremarkable number.",
+      "9999999999 is a number for which we're missing a fact (submit one to numbersapi at google mail!).",
+    ];
+    const sentence = getDefaultMsg({
+      number: 9999999999,
+      type: "math",
+      options: {},
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:257]: same here, can you remove this line?

File: __tests__/unit/fact.test.js [hunk 16]
```diff
@@ -0,0 +1,415 @@
...
+    });
+    expect(mathMsgs.includes(sentence)).toEqual(true);
+  });
+
+  test("default msg for type:'date' with no fact", function () {
+    const sentence = getDefaultMsg({
+      number: 1 / 1,
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:264]: same comment as above, maybe i don't quite understand how the API works with dates

File: __tests__/unit/fact.test.js [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+      type: "date",
+      options: {},
```
Comment (mmmaaatttttt) [inline __tests__/unit/fact.test.js:266]: can you remove this line?

File: models/data.js [hunk 18]
```diff
@@ -47,6 +47,7 @@ function reader_norm(out, pathname, callback) {
       );
       return;
     }
+
```
Comment (mmmaaatttttt) [inline models/data.js:50]: can you revert the changes in this file? i don't think we need these newlines.

File: models/fact.js [hunk 19]
```diff
@@ -2,40 +2,29 @@ var _ = require("underscore");
...
+    var valid_keys = _.filter(dataKeys[options.type], function (element) {
       return element >= min && element <= max;
     });
+
     return utils.randomChoice(valid_keys);
   }
 }
 
-function getSentence(wantFragment, number, type, data) {
+function getSentence({ wantFragment, number, type, data }) {
```
Comment (mmmaaatttttt) [inline models/fact.js:27]: yeah with this function signature `wantFragment` will be undefined if it's not provided, so you should never need to explicitly pass `wantFragment: undefined` in the tests.

File: models/fact.js [hunk 20]
```diff
@@ -55,12 +44,12 @@ function getSentence(wantFragment, number, type, data) {
   return prefix + " " + text + ".";
 }
 
-function getDefaultMsg(number, type, options) {
+function getDefaultMsg({ number, type, options }) {
```
Comment (mmmaaatttttt) [inline models/fact.js:47]:
```suggestion
function getDefaultMsg({ number, type, options = {} }) {
```

if you provide a default value for `options`, then you should also never need to set `options: {}` in the tests either.

File: __tests__/unit/fact.test.js (hunk 5 above)
Comment (mmmaaatttttt) [reply __tests__/unit/fact.test.js:114]: i hope this will just work, since `getSentence` just checks whether `wantFragment` is undefined or not.

File: models/fact.js [hunk 21]
```diff
@@ -71,23 +60,28 @@ function getDefaultMsg(number, type, options) {
...
   }[type];
 
   var data = {
-    text: utils.randomChoice(defaultMsgs)
+    text: utils.randomChoice(defaultMsgs),
   };
 
-  return getSentence(options.fragment, number, type, data);
+  return getSentence({
+    wantFragment: options.fragment,
```
Comment (mmmaaatttttt) [inline models/fact.js:72]: yea, in this case i think the value of `wantFragment` will be a boolean or undefined, not an object. i could be wrong though. in any event, i *think* this means you should just be able to set `wantFragment: true` in those earlier tests.

**Review by mmmaaatttttt — CHANGES_REQUESTED** [review]
@kathyn262 @WinnieGotYu - this looks great!! just some very minor cleanup and then i'll merge this, let me know if you have any questions.

File: __tests__/unit/fact.test.js (hunk 14 above)
Comment (WinnieGotYu) [reply __tests__/unit/fact.test.js:236]: @mmmaaatttttt when `options : { }` is removed, `TypeError: Cannot read property 'fragment' of undefined` occurs. 

The return statement is looking for an `options` object: 
```
return getSentence({
    wantFragment: options.fragment,
    number: number,
    type: type,
    data: data,
  });
```

I'm leaning towards keeping the `options : { }` in the tests ?

Comment (WinnieGotYu) [reply __tests__/unit/fact.test.js:236]: Or another option would be to modify `function getDefaultMsg({ number, type, options })` to `function getDefaultMsg({ number, type, options = {} })` ?

Comment (WinnieGotYu) [reply __tests__/unit/fact.test.js:236]: Sorry for the spam, I think I'll go with options b and modify the getDefaultMsg() to have a default empty object as options if no options is passed in

**Comment by WinnieGotYu** [comment]
@mmmaaatttttt, @kathyn262 and I have completed the suggested updates. 🤞

**Comment by WinnieGotYu** [comment]
we also changed the file inputs for reader_norm and reader_manual in data.js it was pointing to a wrong path earlier

File: models/fact.js [hunk 22]
```diff
@@ -1,41 +1,31 @@
 var _ = require("underscore");
 var data = require("./data.js");
 var utils = require("../public/js/shared_utils.js");
+const { Console } = require("console");
```
Comment (mmmaaatttttt) [inline models/fact.js:4]:
```suggestion
```

this is an unused import, let's remove.

**Review by mmmaaatttttt — CHANGES_REQUESTED** [review]

**Review by mmmaaatttttt — APPROVED** [review]
